domain/network/state/subnet.go

```diff
@@ -80,103 +116,135 @@ VALUES (?, ?)`
+		}
+		// Add the association of the underlay and the newly
+		// created subnet to the associations table.
 		if _, err := tx.ExecContext(
 			ctx,
-			insertSubnetStmt,
-			uuid.String(),
-			cidr,
-			VLANTag,
-			spaceUUIDValue,
-			subnetType,
+			insertSubnetAssociationStmt,
+			uuid,
+			underlaySubnetUUID,
 		); err != nil {
 			return errors.Trace(err)
 		}
-
-		if subnetType == subnetTypeFanOverlaySegment {
-			// Retrieve the underlay subnet uuid.
-			var underlaySubnetUUID string
-			row := tx.QueryRowContext(ctx, retrieveUnderlaySubnetUUIDStmt, fanInfo.FanLocalUnderlay)
-			if err := row.Scan(&underlaySubnetUUID); err != nil {
+	}
+	// Add the subnet uuid to the provider ids table.
+	if _, err := tx.ExecContext(
+		ctx,
+		insertSubnetProviderIDStmt,
+		providerID,
+		uuid,
+	); err != nil {
+		return errors.Trace(err)
+	}
+	// Add the subnet and provider network uuids to the
+	// provider_network_subnet table.
+	if _, err := tx.ExecContext(
+		ctx,
+		insertSubnetProviderNetworkIDStmt,
+		pnUUID.String(),
+		providerNetworkID,
+	); err != nil {
+		return errors.Trace(err)
+	}
+	// Insert the providerNetworkUUID into provider network to
+	// subnets mapping table.
+	if _, err := tx.ExecContext(
+		ctx,
+		insertSubnetProviderNetworkSubnetStmt,
+		pnUUID.String(),
+		uuid,
+	); err != nil {
+		return errors.Trace(err)
+	}
+	for _, az := range availabilityZones {
+		// Retrieve the availability zone.
+		var azUUIDStr string
+		row := tx.QueryRowContext(ctx, retrieveAvailabilityZoneStmt, az)
+		err := row.Scan(&azUUIDStr)
+		if err != nil && err != sql.ErrNoRows {
+			return errors.Trace(err)
+		}
+		// If it doesn't exist, add the availability zone.
+		if err == sql.ErrNoRows {
+			azUUID, err := utils.NewUUID()
+			if err != nil {
 				return errors.Trace(err)
 			}
-			// Add the association of the underlay and the newly
-			// created subnet to the associations table.
+			azUUIDStr = azUUID.String()
 			if _, err := tx.ExecContext(
 				ctx,
-				insertSubnetAssociationStmt,
-				uuid.String(),
-				underlaySubnetUUID,
+				insertAvailabilityZoneStmt,
+				azUUIDStr,
+				az,
 			); err != nil {
 				return errors.Trace(err)
 			}
 		}
-		// Add the subnet uuid to the provider ids table.
+		// Add the subnet id along with the az uuid into the
+		// availability_zone_subnet mapping table.
 		if _, err := tx.ExecContext(
 			ctx,
-			insertSubnetProviderIDStmt,
-			providerID,
-			uuid.String(),
+			insertAvailabilityZoneSubnetStmt,
+			azUUIDStr,
+			uuid,
 		); err != nil {
 			return errors.Trace(err)
 		}
-		// Add the subnet and provider network uuids to the
-		// provider_network_subnet table.
-		if _, err := tx.ExecContext(
+	}
+	return nil
+}
+
+// AddSubnet creates a subnet.
+func (st *State) AddSubnet(
+	ctx context.Context,
+	uuid string,
+	cidr string,
+	providerID network.Id,
+	providerNetworkID network.Id,
+	VLANTag int,
+	availabilityZones []string,
+	spaceUUID string,
+	fanInfo *network.FanCIDRs,
+) error {
```

Comment:
There are too many parameters here. Make an intermediate type, you can then pass that into addSubnet as well.